src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5570,36 +5511,37 @@ private static Task InternalWhenAll(Task[] tasks)
         // which involves several allocations, with this logic:
         //      return new WhenAllPromise(tasksCopy);
         // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
             /// <summary>The number of tasks remaining to complete.</summary>
             private int m_count;
+            /// <summary>The tasks that faulted.</summary>
+            private ConcurrentQueue<Task>? m_faultedTasks;
+            /// <summary>The first task that got canceled.</summary>
+            private Task? m_canceledTask;
 
-            internal WhenAllPromise(Task[] tasks)
+            internal WhenAllPromise(IEnumerable<Task?>? tasks)
             {
-                Debug.Assert(tasks != null, "Expected a non-null task array");
-                Debug.Assert(tasks.Length > 0, "Expected a non-zero length task array");
+                if (tasks == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
 
                 if (TplEventSource.Log.IsEnabled())
                     TplEventSource.Log.TraceOperationBegin(this.Id, "Task.WhenAll", 0);
 
                 if (s_asyncDebuggingEnabled)
                     AddToActiveTasks(this);
 
-                m_tasks = tasks;
-                m_count = tasks.Length;
-
-                foreach (Task task in tasks)
+                int count = 0;
+                foreach (Task? task in tasks)
                 {
+                    if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
                     if (task.IsCompleted) this.Invoke(task); // short-circuit the completion action, if possible
                     else task.AddCompletionAction(this); // simple completion action
+                    count += 1;
```

Comment:
This has multiple race conditions. For example, the moment you've hooked up a continuation to a task, it could race to complete such that it's decrementing the count while this is incrementing it, without synchronizaton. And a task could complete and decrement the count to zero even when more tasks remain.

Reply:
The incrementation is done only one line 5542 and I think it avoids these race conditions.

Reply:
Ah, yes, reading on my phone and didn't see the distinction between count and _count.